### PR TITLE
sftp: run private SFTP server based on Docker (atmoz/sftp)

### PR DESCRIPTION
**Pull request by jkowalski**

### CODE REVIEW

File: internal/testutil/dockertestutil.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	c := exec.Command("docker", args...) //nolint:gosec
+
+	var stderr bytes.Buffer
+
+	c.Stderr = &stderr
+
+	out, err := c.Output()
+	if err != nil {
+		// skip or fail hard when running in CI environment.
+		TestSkipUnlessCI(t, "unable to run docker: %v %v (stderr %v)", err, string(out), stderr.String())
```
Comment (julio-lopez) [inline internal/testutil/dockertestutil.go:26]: nit: 
```suggestion
		TestSkipUnlessCI(t, "unable to run docker: %v %s (stderr %v)", err, out, stderr.String())
```

File: internal/testutil/testutil.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+	if len(args) > 0 {
+		msg = fmt.Sprintf(msg, args...)
+	}
+
+	if os.Getenv("CI") != "" {
+		t.Fatal(msg)
+	} else {
+		t.Skip()
+	}
```
Comment (julio-lopez) [inline internal/testutil/testutil.go:23]: nit: Is the intent to pass `msg` to `t.Skip()` as well? If not, why not skip and bail out early when appropriate?


```suggestion
	if os.Getenv("CI") == "" {
		t.Skip()
	} 

	if len(args) > 0 {
		msg = fmt.Sprintf(msg, args...)
	}

	t.Fatal(msg)
```

File: repo/blob/sftp/sftp_storage_test.go [hunk 3]
```diff
@@ -4,23 +4,130 @@ import (
...
+const (
+	dockerImage  = "atmoz/sftp"
+	dialTimeout  = 10 * time.Second
+	sftpUsername = "foo"
+)
+
+func mustGetLocalTmpDir(t *testing.T) string {
+	t.Helper()
+
+	tmpDir, err := ioutil.TempDir(".", ".creds")
```
Comment (julio-lopez) [inline repo/blob/sftp/sftp_storage_test.go:32]: Why doesn't `t.TempDir()` work here?

File: repo/blob/sftp/sftp_storage_test.go [hunk 4]
```diff
@@ -4,23 +4,130 @@ import (
...
+
+func mustRunOrSkip(t *testing.T, cmd string, args ...string) []byte {
+	t.Helper()
+
+	o, err := exec.Command(cmd, args...).CombinedOutput()
+	if err != nil {
+		t.Fatal(string(o), err)
+	}
+
+	t.Logf("output: %v", string(o))
```
Comment (julio-lopez) [inline repo/blob/sftp/sftp_storage_test.go:57]: nit:
```suggestion
	t.Logf("output: %s", o)
```

File: repo/blob/sftp/sftp_storage_test.go [hunk 5]
```diff
@@ -4,23 +4,130 @@ import (
...
+		ioutil.WriteFile(knownHostsFile, knownHostsData, 0600)
+
+		t.Logf("SFTP server OK on host:%q port:%v. Known hosts file: %v", host, port, knownHostsFile)
+
+		return
+	}
+
+	t.Skipf("SFTP server did not start!")
+
+	return //nolint:nakedret
```
Comment (julio-lopez) [inline repo/blob/sftp/sftp_storage_test.go:112]: What is the effect of removing this return?

**Review by julio-lopez — APPROVED** [review]

File: repo/blob/sftp/sftp_storage_test.go (hunk 3 above)
Comment (jkowalski) [reply repo/blob/sftp/sftp_storage_test.go:32]: for some reason the permissions/mountpoints of `t.TempDir()` are somehow messing with docker. at least did not work for me on a mac.

File: repo/blob/sftp/sftp_storage_test.go (hunk 4 above)
Comment (jkowalski) [reply repo/blob/sftp/sftp_storage_test.go:57]: does it work, though? "o" is []byte

File: repo/blob/sftp/sftp_storage_test.go (hunk 5 above)
Comment (jkowalski) [reply repo/blob/sftp/sftp_storage_test.go:112]: it does not compile.

File: internal/testutil/testutil.go (hunk 2 above)
Comment (jkowalski) [reply internal/testutil/testutil.go]: fixed

File: repo/blob/sftp/sftp_storage_test.go (hunk 4 above)
Comment (jkowalski) [reply repo/blob/sftp/sftp_storage_test.go:57]: it does, fixed.

File: internal/testutil/dockertestutil.go (hunk 1 above)
Comment (jkowalski) [reply internal/testutil/dockertestutil.go]: done

File: repo/blob/sftp/sftp_storage_test.go (hunk 3 above)
Comment (julio-lopez) [reply repo/blob/sftp/sftp_storage_test.go:32]: got it, thanks the for explanation. It'd be good to include a comment as others or our future selves may wonder the same.

File: repo/blob/sftp/sftp_storage_test.go (hunk 5 above)
Comment (julio-lopez) [reply repo/blob/sftp/sftp_storage_test.go:112]: Oh, I see why.

**Pull request merged**